### DISCUSSION TITLE
API Changed // Setting cue_out not working anymore

### DISCUSSION THREAD

**Discussion by BornKillaz** [AzuraCast/AzuraCast#7234] [Q&A] [answered]
Hi.

**Rolling Release #f70c987**

In my custom DJ's panel I set the **cue_out** value on pre-recorded shows using AzuraCast API. 

However I notice that this value is no longer being set, but still returns success:
`Array ( [success] => 1 [message] => Record updated successfully. [formatted_message] => Record updated successfully. )`

    public function updateMediaFileDetails($mediaId, array $updateData) {

        // Check if mediaId is an integer
        if (filter_var($mediaId, FILTER_VALIDATE_INT) === false) {

            error_log("Invalid media ID: {$mediaId}. Must be a non-decimal number.");

            return false;
        }

        // Construct the endpoint to update the file
        $endpoint = "/station/{$this->stationId}/file/{$mediaId}";

        try {
            // Make the API request to update the file
            $response = $this->apiRequest($endpoint, 'PUT', $updateData);

	   print_r($response); // DEBUG

            // Check the response for success indicator
            return isset($response['success']) && $response['success'] == 1;
        } catch (Exception $e) {
            // Log the error and throw
            error_log("Error updating media file: " . $e->getMessage());
            throw $e;
        }
    }
    

```
 $updateData = ['cue_out' => 3600];

// Update the media file details with the new cue_out value
if (!$panel->updateMediaFileDetails(4127, $updateData)) {
            
    exit('Could not update');
	
}
```

I see that the current json is like this, and there used to be a cue_out field but not anymore:
```
{
  "text": "Chet Porter - Aluko River",
  "artist": "Chet Porter",
  "title": "Aluko River",
  "album": "Moving Castle",
  "genre": "Rock",
  "isrc": "US28E1600021",
  "lyrics": "",
  "links": {
    "additionalProp1": "string",
    "additionalProp2": "string",
    "additionalProp3": "string"
  },
  "id": 1,
  "unique_id": "69b536afc7ebbf16457b8645",
  "song_id": "9f33bbc912c19603e51be8e0987d076b",
  "art": "https://picsum.photos/1200/1200",
  "path": "test.mp3",
  "mtime": 1609480800,
  "art_updated_at": 1609480800,
  "length": 240,
  "length_text": "4:00",
  "custom_fields": [
    "custom_field_value"
  ],
  "extra_metadata": [
    "string"
  ],
  "playlists": [
    "string"
  ]
}
```

How can I now set cue_out value? Setting it in the new  "extra_metadata" field **does not work**:

```
  "extra_metadata": [
    "cue_out": 3600
  ]
```

I can only set liq_cue_out on extra_metadata, **this works**:
```
  "extra_metadata": [
    "liq_cue_out": 3600
  ]
```

However, I usually set **cue_out** so that it has priority over liq_cue_out and I let AutoCue deal with those liq_xxxx, but even AzuraCast visual editor is now ignoring cue_out and uses liq_cue_out instead.

Also, for example, what would happen if i set **liq_cue_out to 3600** with API and AutoCue set **liq_cross_start_next to 3994.7**? 

Thank you.

**Comment by Moonbase59** [comment]
The API has changed, study the code.

**Reply by BornKillaz** [reply]
Thank you very much for the kind help on this subject.

I already updated my code in a way it makes sense to me.

**Reply by Moonbase59** [reply, marked as answer]
Just try not to mess up things… ;-)

I can’t and won’t support any third-party software, but I just had a quick look for you:

```
  "extra_metadata": {
    "liq_fade_in": 0.1,
    "liq_fade_out": 0.1
  },
```

`extra_metadata` is a _dictionary_, not a list (!). You’d have to add/modify the `liq_cue_out` entry, which is a float:

```
  "extra_metadata": {
    "liq_fade_in": 0.1,
    "liq_fade_out": 0.1,
    "liq_cue_out": 3600.0
  },
```

Note that’s JSON, so watch your commata! (No extra comma at end of dict.)

For easy testing, you can use your Swagger at `https://yourstation.com/api`.

> Also, for example, what would happen if i set liq_cue_out to 3600 with API and AutoCue set liq_cross_start_next to 3994.7?

_That’s_ why I usually recommend _not_ to meddle with the values, since some are dependent on others. In your case (just modifying the cue-out point), Autocue will try to correct the error, taking into account a fade-out duration, if set. So it would effectively set the `liq_cross_start_next` to `liq_cue_out` - `liq_fade_out`. And give a message in the log as to the change.

```
[autocue.cue_file:3] Given liq_cross_start_next (3994.7 s) > cue-out point (3600. s), set to 3595.  s.
```

**Reply by BornKillaz** [reply]
Thank you, those were great tips and I manage to properly update my logics.

**Reply by Moonbase59** [reply]
Good to hear you succeeded!

The background of this API change is to be able to store Autocue (and possibly other) "extra" metadata under their original tag/metadata names. This makes it easier to diagnose issues (same name you as see in the logs/file tags) and will eventually be the means to feed back data from Autocue into the Visual Cue Editor.

**Reply by BornKillaz** [reply]
Just one more doubt, if you can help, what if I want a more smooth fade in/fade out? Let's say, 3 secunds overlap, is it safe to do so while defining cue out to 3600.0 for a file that exceeds the 1 hour duration.

**Reply by Moonbase59** [reply]
Well, if you use the API, you could set

```
  "extra_metadata": {
    "liq_fade_in": 0.1,
    "liq_fade_out": 3.0,
    "liq_cue_out": 3600.0
  },
```

And Autocue should be clever enough to adjust:

```
[autocue.cue_file:3] Given liq_cross_start_next (3994.7 s) > cue-out point (3600. s), set to 3597.  s.
```

Or, to make it more correct, calculate and set `liq_cross_start_next`, too:

```
  "extra_metadata": {
    "liq_fade_in": 0.1,
    "liq_fade_out": 3.0,
    "liq_cross_start_next": 3597.0,
    "liq_cue_out": 3600.0
  },
```

(Not tested but should work.)

**Reply by BornKillaz** [reply]
> Well, if you use the API, you could set
> 
> ```
>   "extra_metadata": {
>     "liq_fade_in": 0.1,
>     "liq_fade_out": 3.0,
>     "liq_cue_out": 3600.0
>   },
> ```
> 
> And Autocue should be clever enough to adjust:
> 
> ```
> [autocue.cue_file:3] Given liq_cross_start_next (3994.7 s) > cue-out point (3600. s), set to 3597.  s.
> ```
> 
> Or, to make it more correct, calculate and set `liq_cross_start_next`, too:
> 
> ```
>   "extra_metadata": {
>     "liq_fade_in": 0.1,
>     "liq_fade_out": 3.0,
>     "liq_cross_start_next": 3597.0,
>     "liq_cue_out": 3600.0
>   },
> ```
> 
> (Not tested but should work.)

Amazing, I'll do just that! Thank you.
